data/coinductive.lean

```diff
@@ -0,0 +1,712 @@
+
+Work in progress
+-/
+import data.pfun
+
+import data.stream
+
+universes u v w
+
+open nat function roption list
+
+variables {α : Type u}
+variables (β : α → Type v)
+
+local prefix `♯`:0 := cast (by simp [*] <|> cc <|> solve_by_elim)
+
+local attribute [instance, priority 0] classical.prop_decidable
+
+namespace coinduction.approx
+
+/-
+coinductive ind {α : Type u} (β : α → Type v) : Type (max u v)
+| intro : ∀ a, (β a → ind) → ind
+-/
+
+inductive cofix_a : ℕ → Type (max u v)
+| continue : cofix_a 0
+| intro {n} : ∀ a, (β a → cofix_a n) → cofix_a (succ n)
+
+variables {β}
+
+def head' : Π {n}, cofix_a β (succ n) → α
+ | n (cofix_a.intro i _) := i
+
+def children' : Π {n} (x : cofix_a β (succ n)), (β (head' x) → cofix_a β n)
+ | n (cofix_a.intro _ f) := f
+
+def agree
+: ∀ {n : ℕ}, cofix_a β n → cofix_a β (n+1) → Prop
+ | 0 continue _ := true
+ | n (cofix_a.intro x y) (cofix_a.intro x' y') :=
+   x = x' ∧ ∀ i j : β _, i == j → agree (y i) (y' j)
+
+def all_agree (x : Π n, cofix_a β n) :=
+∀ n, agree (x n) (x (succ n))
+
+@[simp]
+lemma agree_trival {x : cofix_a β 0} {y : cofix_a β 1}
+: agree x y :=
+by { cases x, trivial }
+
+lemma agree_def {n : ℕ} (x : cofix_a β (succ n)) (y : cofix_a β (succ n+1))
+  (h₀ : head' x = head' y)
+  (h₁ : ∀ (i : β _) (j : β _), i == j → agree (children' x i) (children' y j))
+: agree x y :=
+begin
+  cases x, cases y,
+  unfold agree,
+  cases h₀,
+  existsi rfl,
+  unfold children' at h₁,
+  intro, apply h₁,
+end
+
+lemma agree_children {n : ℕ} (x : cofix_a β (succ n)) (y : cofix_a β (succ n+1))
+  {i j}
+  (h₀ : i == j)
+  (h₁ : agree x y)
+: agree (children' x i) (children' y j) :=
+begin
+  cases x, cases y,
+  unfold agree at h₁,
+  cases h₁ with h h₁, subst x_a,
+  unfold children',
+  cases h₀, apply h₁,
+  assumption,
+end
+
+def truncate {α : Type u} {β : α → Type v}
+: ∀ {n : ℕ}, cofix_a β (n+1) → cofix_a β n
+ | 0 (cofix_a.intro _ _) := cofix_a.continue _
+ | (succ n) (cofix_a.intro i f) := cofix_a.intro i $ truncate ∘ f
+
+lemma truncate_eq_of_agree {α : Type u} {β : α → Type v} {n : ℕ}
+  (x : cofix_a β n)
+  (y : cofix_a β (succ n))
+  (h : agree x y)
+: truncate y = x :=
+begin
+  revert x y,
+  induction n
+  ; intros x y
+  ; cases x ; cases y,
+  { intro h', refl },
+  { simp [agree,truncate,exists_imp_distrib],
+    introv h₀ h₁,
+    subst x_a, split, refl,
+    apply heq_of_eq, funext y, unfold comp,
+    apply n_ih,
+    apply h₁, refl }
+end
+
+variables {X : Type w}
+variables (f : X → Σ y, β y → X)
+
+def s_corec : Π (i : X) n, cofix_a β n
+ | _ 0 := cofix_a.continue _
+ | j (succ n) :=
+   let ⟨y,g⟩ := f j in
+   cofix_a.intro y (λ i, s_corec (g i) _)
+
+lemma P_corec (i : X) (n : ℕ) : agree (s_corec f i n) (s_corec f i (succ n)) :=
+begin
+  revert i,
+  induction n with n ; intro i,
+  trivial,
+  cases h : f i with y g,
+  simp [s_corec,h,s_corec._match_1,agree] at ⊢ n_ih,
+  introv h',
+  cases h',
+  apply n_ih,
+end
+
+abbreviation path' (β : α → Type v) := list (Σ i, β i)
```

Comment:
AFAIK, the `abbreviation` command is not intended for general usage.  Its only purpose is to define `id_rhs`.  The difference compared to `def` is that it sets a special reducibility hint for the kernel type-checker.

Reply:
Thanks for the note! I thought I read a comment from Leo about `abbreviation` removing the need to ever use `reducible` but I can't find it anymore. In any case, this definition works even without `reducible` so I'll just make it plain.